### ISSUE TITLE
Recurrent Batch Normalization

### ISSUE THREAD

**Issue by lakehanne** [Element-Research/rnn#252]
I want to implement the [recurrent batch normalization](http://arxiv.org/pdf/1603.09025v4.pdf) as proposed in this paper: http://arxiv.org/pdf/1603.09025v4.pdf. But before I go ahead, I want to be sure you do not have a similar implementation already.

**Comment by JoostvDoorn** [comment]
I guess no one is implementing this right now. You can also have a look here: https://github.com/iassael/torch-bnlstm/blob/master/LSTM.lua

**Comment by nicholas-leonard** [comment]
@lakehanne Feel free to implement it for SeqLSTM.

**Comment by jnhwkim** [comment]
@lakehanne If you don't mind, you can reference my attempt (https://github.com/jnhwkim/rnn/tree/bn). I tried to implement the recurrent batch normalization on GRUs, but I failed and I gave up, though. I didn't try for LSTMs. I think you can easily control shared parameters using `dpnn_parameters` and `dpnn_gradParameters`. See https://github.com/Element-Research/dpnn/blob/5ec0d1ad39a8aa234a343de034d9a5003d097c67/Module.lua#L37. 
@nicholas-leonard, my thought was right?

**Comment by lakehanne** [comment]
Thanks @JoostvDoorn for the link. That was very helpful. 

Thanks to you @jnhwkim too.

I modified the `FastLSTM` class to use the `recurrent batch normalization` in the paper above. It's not been massively unit tested yet but I would think it works. I have tested it on a couple of datasets of mine and it seems ok. Would appreciate what you guys think of it. Code [here](https://gist.github.com/lakehanne/f59dabce2c2d93376069f6c1733dc6ea).

I am working on adapting this for the `LSTM` module of @nicholas-leonard but I have been constrained with time lately. I will conclude that soon and let you know how it goes.

**Comment by jnhwkim** [comment]
@lakehanne It seems you are editing, now. Few comments:
* Recommend to match filename with classname
* gammas and betas are shared parameters, while learning averages of mean and std of BatchNormalization should be separate at each timestep.

**Comment by lakehanne** [comment]
Thanks @jnhwkim . A few comments on your unordered lists observations here:

+ Yeah, I figured. I have this modified [here](https://github.com/lakehanne/FARNN/blob/repeater/utils/BNFastLSTM.lua) now

- This is done in the `nn.BatchNormalization` class based on [Ioffe and Szegedy's](https://github.com/torch/nn/blob/master/BatchNormalization.lua) [work](http://arxiv.org/pdf/1502.03167.pdf)? Only difference I seem to notice is that the [RBN authors](http://arxiv.org/pdf/1603.09025v4.pdf) took the square root of the denominator whereas Ioffe and Szegedy did not. I do not understand the rationale behind the square root though. You are saying we need to separate the batch normalization from the way I am currently handling them? 

- I will look into that.

Thanks for your observations.

**Comment by nicholas-leonard** [comment]
@jnhwkim Yes your can control shared parameters via `dpnn_parameters` and `dpnn_gradParameters`. Hope you will be sending a PR with your cool new GRU additions :)

@lakehanne Yes please send a PR. Concerning your implementation, it would be awesome if you could just modify FastLSTM (instead of subclassing into BNFastLSTM) to add batch-normalization after construction. So use case would be : 
```lua
lstm = nn.FastLSTM(inputSize, outputSize)
lstm:batchNormalization(eps, momentum, affine)
```

That would be the sexiest solution.  I am not an nngraph expert, but I think you could modify the recurrentModule to replace the input Identities with batchnormalization.

**Comment by jnhwkim** [comment]
@nicholas-leonard Sadly, I failed to validate BNGRU. No better than simple dropout in validation metric. The project is suspended.

**Comment by nicholas-leonard** [comment]
@jnhwkim Sounds good. Thanks for looking into this :)

**Comment by viorik** [comment]
Hi Guys,
I am looking into implementing (spatial) batch normalisation for [ConvLSTM](https://github.com/viorik/ConvLSTM), and your implementation would be a good starting point. Just a small (possibly silly) question: in your implementation I see the original Linear and LinearNoBias layers with BN on top. This means that the bias of the Linear layer is under BN, but in the [paper](https://arxiv.org/pdf/1603.09025v4.pdf), at p.4 eq. 6, it appears to be separate from the BN functions? Is that a typo in the paper or maybe I misunderstood something? 
Thanks in advance for any clarification.

**Comment by lakehanne** [comment]
Thanks @viorik for your observations. @nicholas-Leonard, she seems to be raising a valid point. What do you think? Shouldn't 'i2g' be using a 'LinearNoBias' class?

**Comment by nicholas-leonard** [comment]
@viorik I see the difference between the rnn implementation and the paper. But I really don't think it would change much to performance whether bias is before or after BN. My hunch is based on the fact that BN is used just fine after biases in normal non-recurrent use-cases.

@lakehanne Feel free to send a PR to modify the implementation so that it fits the paper.

**Comment by nicholas-leonard** [comment]
@viorik fixed here https://github.com/Element-Research/rnn/commit/fb6aaea2efb6b9ddc9449779d72318460fcc613d. Thanks for bringing this up.

**Comment by viorik** [comment]
Hi Guys,
Sorry for coming back to a closed issue, but I would have a question. 
The statistics computed during training are independent for each time step. At evaluation time, we are supposed to use the average statistics over the training set, which are stored internally in `running_mean` and `running_var` of the `BatchNormalization` layer. But since we have separate means and var for each step, at evaluation time, which ones does your implementation use? I implemented batchNorm for my ConvLSTM layer, following your implementation, but it seems to me that in my implementation, at evaluation time, I end up using the mean and var from the first time step, since `recurrentModule` is `sharedClone[1]`. And the results don't look good in my case. Maybe we should use the average over all time steps. Any thought on this? Or maybe I am wrong on this and there is something else going wrong in my code.
Any help appreciated.

**Comment by nicholas-leonard** [comment]
@viorik Sorry for delay. For sharing these things between different time-steps (via sharedClone), simply make sure they are part of the [BatchNormalization.dpnn_parameters](https://github.com/Element-Research/dpnn/blob/master/BatchNormalization.lua#L15-L18). I think they already are as this issue came up in the past. Correct me if I am wrong?

**Comment by viorik** [comment]
@nicholas-leonard thank you for your reply. I guess I need to do that for SpatialBatchNormalization, since I use the spatial version for my ConvLSTM. Hope that fixes the issue. Thank you again.
